### ISSUE TITLE
Right-click pipes holding chipsets to upgrade them.

### ISSUE THREAD

**Issue by HaniiPuppy** [RS485/LogisticsPipes#637]
Where logistics pipes can be upgraded by just crafting them with a chipset, it would be good to be able to right-click them holding a chipset to upgrade them, rather than having to break the block and craft them in a crafting grid, where you'd have to re-add any plugs, façades, modules, etc. and anything connected to the rest of the pipe system becomes disconnected while you do it.

I think this would be primarily ideal for upgrading mk4 chassis pipes to mk5 (mk4 + diamond chipset), it's a pain placing the modules back, and the covers when you do things like hide the piping in the walls and floors.

**Comment by Sidorion** [comment]
I second this. 
Especially crafting pipes are a pain in the butt to upgrade since you have to reset the recipe, the satellites, liquid crafting upgrades and so on.

**Comment by hron84** [comment]
+1 vote

**Comment by SonOfTheStars** [comment]
+1000

**Comment by SonOfTheStars** [comment]
Ill give this a workabout

**Comment by davboecki** [comment]
I don't like the idea of simply clicking with an chipset on a pipe to upgrade it. But the general idea of upgradable pipes is not bad. I currently think about making it part of the pipe controller. But than it will require not only the chip to upgrade but one additional item and some power. But I havn't decided anything yet.

**Comment by hron84** [comment]
@davboecki or a handy chargeable upgrade device (think to electric wrench in IC2 but only for upgrades)

**Comment by SonOfTheStars** [comment]
@davboecki and I discussed a bit about maybe adding a Tab to the Pipe Controller. Adding Certain parts into the Pipe and Catalysing the Reaction with an Enderpearl. Upon the reactions start, the Pipe Disconnects and starts the Upgrade Process.
While the Pipe Upgrades it emits Sparkles, wich peak upon Completion.

After the Process, the Upgraded Pipe gets Re-Added to the Network.

**Comment by bziemons** [comment]
I think the network (and item flow) should not be interrupted when you upgrade a pipe.
The idea sounds nice, but it would be very annoying.

**Comment by SonOfTheStars** [comment]
@theZorro266 Well, There needs to be a Penalty for replacing something in place. Wich System can handle an Augmentation while its running?

Plus, mainly people do have functional Pipes at the outer edges of their Systems. So the effect would be minimal.

Additionaly upgrade times Variates as the upgrades go higher.

**Comment by SonOfTheStars** [comment]
I would be very pleased if it would work that way.

Im okay with no disconnect, but that will require MUCH longer upgrade times

**Comment by davboecki** [comment]
Yea disconnectiong would probably be bad in some situations. Rather disable the pipes functionality while it upgrades and require an ender pearl next to the crafting items to upgrade the pipe. And consume power while the upgrade is in progress.

**Comment by bziemons** [comment]
Hmm, an ender pearl sounds quite expensive when you do not have mods farming them for you.
Two reasons why I would upgrade a pipe inplace are: Keep the settings and do not interrupt the network.

A penalty in time when upgrading inplace sounds okay, because instant would be another "unnecessary" positive side-effect (also that penalty should be reasonable. I am thinking of around 5 seconds).

Making the cost of an inplace upgrade in terms of material higher sounds okay, but it needs to be something that is not too hard to get when you just play with e.g. BuildCraft and Logistics Pipes. I think of something like the upgrade slots for JABBA. They are not too expensive, but an extra item you need to craft. I would like to see something like that. Also keep in mind that "upgrading takes power" is already an extra penalty.

**Comment by SonOfTheStars** [comment]
@theZorro266 LP allready uses far less power than anything else, even with a big busy System. The powercosts should be quite big. My current System is eating far less RF/tick than a single upgraded Reactant Dynamo can produce! Wich is 160/tick

**Comment by Silwing** [comment]
As soon as you get to the end and defeat the dragon ender pearls is a pretty cheap resource even without many mods. And before then the difficult thing is to find the endermen in the overworld, since you can always make a 2-block shelter and kill them off even with no armor and a cheap sword.

**Comment by hron84** [comment]
@Silwing Currently LP is playable on Peaceful difficulty too it would be break if you need ender pearl or even ender eye. I agree with the argument we need penality however we should not break the playability without any alternative. All resources that needed for LP is obtainable in Peaceful without any mob kill. 

Just my $0.02.

**Comment by hron84** [comment]
@Bitterholz energy cost is OK, however it should not bigger than the Junction table can hold because it can cause problems. 

Alternative is a configurable way: use ender pearl or some expensive upgrade holder (i think to something like storage cells and cell housing in AE) and take instant upgrade or upgrade and take big amount of power or both.

**Comment by bziemons** [comment]
@Bitterholz My argument was not to remove power usage, but to keep the cost reasonable.
What actually can cost a lot of energy is crafting or having a lot of items travelling all the time, but many mods do not have complicated recipes. Making one HV solar array e.g. drains one power junction pretty fast. Of course I don't think other mods should change to complicated recipes, but my point is that our difficulty also relies on other mods. I do not think that Logistics Pipes is that easy, also to mention that people need to make up a good way to plan their entire network to actually make it use the least energy.

@Silwing Yes, but I have a Logistics Pipes network and most of the times a well-built network before I even go to the end. I don't say it is wrong what you say, but as @hron84 said: Peaceful mode.

And if we make it configurable (what is likely to happen, if quite a few seem to want the ender pearls as a cost), we should force the non-enderpearl recipe in peaceful mode. Also we should have a look at other recipes using ender pearls ^^

**Comment by hron84** [comment]
As I saw most people have a working LP system (possibly with Ender Storage chests and pouches) before entering any place like End so @theZorro266 's behavior is not unique.

**Comment by davboecki** [comment]
We are talking about a new feature here. So it wouldn't destroy gameplay even for lower levels if you would need an ender pearl to upgrade a pipe in place. You can still upgarde the pipe by breaking and replacing it if you don't want to pay the price. But in general I see the disadvantage of requiring an ender pearl to upgrade the pipe. Although it would in my oppinion be the perfect item as an inplace upgrade catalyst because an in place upgrade is like teleporting a new pipe into the old pipes place.

**Comment by hron84** [comment]
@davboecki okay, but your argument can be easily attacked if you think to many mod that provides upgrades without requiring ender pearl. Think to Thermal <del>Expansion</del> Dynamics, JABBA, Storage Drawers or even LP as we have sneaky upgrades and these mates and they're not requiring an ender pearl (or not all requires it) but provides a somehow upgraded pipes (even if the block itself is not replaced). So the question is we want to break a tradition to make a penality or we will follow the tradition and figure out a high but acceptable price that can be achieved even in Peaceful.

Even if I say things I do not know how we can do it. The energy cost can be easy to implement but I dunno it's a quite high price for replacing pipes. Ender pearls can be a solution if we can provide a Peaceful recipe to replace pearls with diamonds. We can also require a lot of crafting for these upgrades. There is many-many ways. I just make my 2 cents as I'm a Peaceful player - blame me for it.

**Comment by SonOfTheStars** [comment]
@hron84, The basic jist behind the Enderpearl is:
Youre changeing something that is presistent in the World.

Thermal Expansion gets pretty costly on the Augments very Quick.
I don´t know many ppl who Play Peacefull mode. But youre right, the case is there. BVut Peacefull is detectable, so you technically COULD do it such that as soon as you enter pecaefull you dont need the Enderpearl.

And i can prove you wrong, LP requires Enderpearls in a few Recipes allready! The BETA recipes in 0.9 do also feature Slimeballs for Sticky Pistons.

@davboecki After all, this is nothing but a Comfort Addition. As you said, you can allways break and replace. I like this feature and it defenitely has to come. Balancing the Costs will be quite something ill gie a thinkabout. Allreay thinking about how to present the results

**Comment by bziemons** [comment]
@Bitterholz "And i can prove you wrong, LP requires Enderpearls in a few Recipes allready!"
No, this does not prove him wrong. He actually makes us give attention to players in peaceful mode and the existing ender pearl recipes have to change.

**Comment by hron84** [comment]
@theZorro266 also, not so much recipe needs ender pearl as i know. I never needed an Ender Pearl for my systems - but I never used for example firewall or satellite pipes.

**Comment by TheLoneWolfling** [comment]
Ender pearls would not be a good idea, in my opinion. They are one of those resources that, depending on what you've done, are either extremely rare or dirt cheap. And unfortunately, aren't really correlated with the rest of what stage of the game you're in, as you can anywhere from push for the end immediately through to being in the extreme late game before you go to the end.

Basically, if you make it use ender pearls, you're saying "stop. Now, before you do this, go to the end and build an enderman farm". If that's what you want to do... Then go ahead. But it may not be the best idea overall to do so.

**Comment by Silwing** [comment]
Bringing up Peaceful Mode is a good thing and I agree that recipes should have a peaceful mode replacements if they require hostile mob drops. That would be the nice and sensible thing to do. But I don't think we should avoid mob drops in general as that also takes something away from players not playing peaceful and that would be a shame in my eyes.

There are many ways to get ender pearls, The End is not the only way. And I also agree that it suits the general feeling of ender pearls to use them for this purpose. You can always break and replace.

**Comment by SonOfTheStars** [comment]
How about a Config option for Peacefull Recipes?
So you can Decide via Config to enable the Peacefull setting.
Also @davboecki I will probably try and Split Beta and Classic Recipes on the 0.9 Branch into 2 Different Classes if this happens

**Comment by hron84** [comment]
I just travelled home yesterday and started thinking on this ender pearl problem. Technically currently there is no balance in many mods around ender pearls vs Peaceful, however there is only one common thing: ender pearl is required only in highest tier enhancements and only if the enhancement **constantly** teleports items/blocks/mobs. A simple upgrade is usually not requires ender pearl except it makes the block/item able to **constantly** teleport items/blocks/mobs.

Let me give an example for it. I mentioned it would be OK for me if there is a upgrade tool (like JABBA's hammer or Forestry soldering iron). That tool can be charged with power (possible penality: 2 or 3 upgrade can completely drain the internal energy buffer)  and its recipe should require ender pearl or ender eye as the tool does not break. But consuming ender pearl without any limit just because it is precious is pointless I think. If we try to consume something really precious we should consume emeralds too that is much more rare in the game. 

In case like this I can agree with using ender pearls or any other mob drops - if it is makes sense. But @Bitterholz 's current idea (use one ender pearl at any upgrade action) does not makes any sense for me from the aspect of balancing. 

(alternatively, we can provide a very precious peaceful recipe for that tool with emeralds and/or diamonds if we really want to balance this part out).

@Silwing breaking and replacing in a big LP system is usually not a good idea. I think this teleporting stuff would be good because it can upgrade your system without rebuilding it twice in a row. If you take closer look to our issue tracker we have serious performance problems about attaching/removing a pipe from/to the system and even if our developer guys can avoid the biggest problems the performance of this action is still much higher than seamlessy upgrading a pipe internally.

**Comment by SonOfTheStars** [comment]
@hron84 The Idea behind the Enderpearl was that the only thing thats capable of changing something in Place without having to (re)move it first Is Ender related stuff.

**Comment by Silwing** [comment]
@hron84 I don't think "other mods do this" is a very good argument in itself. If we want LP to be balanced in and of itself it makes more sense to look only at LP and argument from a LP/Vanilla point of view. That ender pearls are somewhat rare (or at least require more effort than mindless mining does) is not the primary argument as to why they make sense in this recipe from my point of view. I'll refer to @Bitterholz last post. So saying emeralds are more rare/expensive doesn't really add anything to the discussion. If rare was the only criterion we could go for Nether Stars, if you see what I'm getting at.

The real issue is peaceful mode. I think any mod for which it makes sense to play in peaceful mode should provide recipes either by config option or forced when peaceful mode is detected, that substitutes any mob drops with other items. Then peaceful mode does not affect whether an ender pearl can be used in the normal recipe.

**Comment by bziemons** [comment]
I don't get why we are at a replacement process now? I thought this was about upgrading pipes and not replacing them with using ender pearls.

Now I see what @davboecki means with "an in place upgrade is like teleporting a new pipe into the old pipes place" and I do honestly not agree with that. If we are upgrading the pipe (which is not teleporting a new one in place), I can imagine that being done like you keep the pipe frame in place, then "take" off the functional part, upgrade that and put the functional part back on the pipe. That's exactly the point, where I do not (and @hron84 may have seen it similar) need an ender pearl.

If we want to make a feature that actually replaces a pipe in your hands with a pipe on the ground, I would call it replacement process and yes: an ender pearl might make sense in that case.

**Comment by Silwing** [comment]
It does make some sense that an upgrade is not directly a replacement. If you see it like that a rechargeable tool like @hron84 suggested would make sense.

Being able to outright replace any LP pipe with another might also be an interesting feature but I guess that should have its own issue if we want to discuss that.

**Comment by AartBluestoke** [comment]
how about the update process takes a bit and consumes significant power,
during which the pipe is disabled (not removed from the network, just as if
a pipe gate had disabled it), and requires the upgrade tool to have an open
gui for 10 seconds or so. The ender pearl could be be used as a catalyst
for a quick upgrade inplace of the energy.

Perhaps if we make the energy cost 5k x the number of pipes in the network,
and the upgrade time 3 ticks/pipe in the network ? that way for small/early
networks you take the power/time penalty without thinking about it, but for
a large network, you'd want the pearl to save deactivating the network with
a broken pipe?

On Fri, Apr 10, 2015 at 5:14 AM, Randi Katrine Hillerøe <
notifications@github.com> wrote:

> It does make some sense that an upgrade is not directly a replacement. If
> you see it like that a rechargeable tool like @hron84
> https://github.com/hron84 suggested would make sense.
> 
> Being able to outright replace any LP pipe with another might also be an
> interesting feature but I guess that should have its own issue if we want
> to discuss that.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/RS485/LogisticsPipes/issues/637#issuecomment-91331607
> .

## 

“Getting information off the Internet is like taking a drink from a fire
hydrant.”
_– Mitchell Kapor_

**Comment by hron84** [comment]
@AartBluestoke i would prefer to not disable the pipe completelly because it may or may not break the system completely, you can't assume. For example, I have a chassis mk2 pipe that is provides (and accepts) a charcoal into a JABBA barrel but also this pipe is in the line that connects my main base with my storage room. If somehow it becomes disabled completely my base will become offline and it's not so funny.

**Comment by HaniiPuppy** [comment]
@AartBluestoke I don't really see how that's any much better than just removing the pipe and crafting it in the crafting grid then replacing it. The feature is largely one of convenience, which is a wee bit defeated when you make it more inconvenient than the thing It's supposed to be an alternate action for.

**Comment by davboecki** [comment]
@theZorro266 and I discussed this and our current thoughts on this are:
- Disable the pipe while upgrading (not the item routing)
- Upgrading takes 30s (configurable) [No need to stay in the GUI]
- A catalyst is needed to do this, New upgrade item [default] or an ender pearl
- The upgrade is triggered out of a new tab inside the pipe controller
- While the upgrde is running you can't open or use the pipe in any way [Exception is the pipe controller]
- The progress is shown inside the pipe controller
- Upgrading a pipe costs 10 000 LP (configurable) (Doesn't scale with the size of the network) [This would be the wrong feature to start with higher power consumtion based on your network size]

**Comment by hron84** [comment]
@davboecki cool I think it seems like a very good solution. Should the power consumption be configurable?
Also, please make the progress visible in WAILA too if it is possible

**Comment by davboecki** [comment]
Will see about WAILA. Yes the power will be configurable.

**Comment by hron84** [comment]
@davboecki cool, thanks!

**Comment by TheLoneWolfling** [comment]
If you're going to add a feature to remove tedium, at least make it something that's actually better than the alternative. That, to me at least, seems worse than just breaking the pipe and replacing it in almost all cases.

**Comment by SonOfTheStars** [comment]
It still saves you the work of break/replace and re-inserting modules

**Comment by HaniiPuppy** [comment]
It's more the interruption of function and network and the time it takes for each one that's the problem, which takes less than 30 seconds. The current proposal by davboecki doesn't address these problems (in fact, seems to be bending over backwards to avoid addressing them while addressing the symptom of literally breaking, upgrading, and placing) and it feels like the kind of feature that's just never going to be used because it's not really any better, it costs more resources (ender pearls are hard to come by, even in modpacks if they don't have enderman grinders or enderlillies), and the fact it's triggered from within the pipe controller means it's both more inconvenient (especially to what I see as the most likely use scenario, which will be people upgrading a single mk4 chassis to mk5 as required) and obscure.

What if, instead, a "Pipe Upgrader" item were crafted with say, paper and redstone circuits (which, for someone with the ability to make iron, gold, or diamond chipsets, is cheap-as-chips) (or something along those lines), which is then crafted with something like a diamond chipset, a gold chipset, a chipset and gear, etc. which can then be used to right-click on a pipe to upgrade it in the way that crafting it with the items crafted with the pipe upgrader would result, using up the upgrader and costing an amount of LP (say, 10k) in the process?

**Comment by bziemons** [comment]
@TheLoneWolfling @HaniiPuppy 
First of all: thanks for your answers. I still do see advantages against breaking and replacing the pipe and I want to explain them. But please do not see this feature as a pure comfort feature. Somehow a few people here (me included) wants to have a trade off, because this feature would be "too easy" otherwise. If you really want this to be a comfort feature: Change the upgrade time and power usage to 0. You can do that, because quite a few people disagreed at how this feature should be "nerfed".

@TheLoneWolfling Sadly you didn't say what exactly you find annoying about the proposal..

@Bitterholz Keep in mind that this is not only about chassis pipes. The plan is to make any (future) pipe that has several tiers upgradable (e.g. crafting pipe), which is even better \o/

@HaniiPuppy The network is not being interrupted. The time (which is configurable) just tells how long the pipe you are upgrading cannot be used. Items will still be routed through that pipe. Maybe in case of one chassis pipe, which does not have a lot of item traffic, this does not make sense. Yes, replacing it, as by breaking it and placing a new one, makes more sense. But also in the mid-game you may want to upgrade a chassis pipe from Mk4 to Mk5, because you do not have that many resources to "waste" them on another Mk4 you need to make for the Mk5.

This **will** make more sense, when the new crafting recipes are done. We had a few suggestions making each chassis pipe with an higher tier based off the previous tier pipe. That means to craft an extra chassis Mk5, you need to craft Mk1 -> Mk2 -> Mk3 -> Mk4 -> Mk5. Same for the crafting pipes, as they will be based off of each other as well. In the end you will then need all items from the recipe for the next tier to actually upgrade a pipe. You suggested using a chipset, but just a chipset does not make sense, when the crafting recipe to make a pipe of the next tier is more expensive (that especially with crafting pipes). Which brings me to the next point..

We mentioned the ender pearl as an additional cost, but I also think that an ender pearl is too expensive. That is why @davboecki also mentioned a second item especially for upgrading a pipe (and that will be the default in the config). Although the crafting recipe of that item is not determined (suggestions are welcome), it will likely be not as expensive as an ender pearl.

Having the upgrade functionality in the pipe controller made me disagree in the first place, too. I was concerned that the tool gets "too heavy" and results in being not used because it would be "too complicated". @davboecki assured me that this would not happen. Instead I also see this as a good opportunity to make the pipe controller being used more by players. Of course we will try to make the tool and the possibility of upgrading a pipe prominent through the wiki and/or some YouTubers, who make showcases on Logistics Pipes. The pipe controller is an amazing tool and if you haven't used it, you should try it out. I do not think that anyone will upgrade a pipe just because you need an extra tool. Also, in the pipe controller, it will be possible to see the upgrade in progress and the remaining seconds, which is again: cool.

For those reasons I would like to not create a new item to upgrade pipes. In the end that would only save you two clicks, but it would have less features (like progress) and cost more.

---

I wrote this huge comment, because different opinions are important. I do think discussions make sense and this whole issue actually shows a good parallelism to software engineering in the real world. If anyone is still not convinced, please give us your comment, but let me sum everything up first:

Upgrading a pipe costs at least a pipe controller and an extra item (like the upgrade item **or** an ender pearl). You are free to configure the time that it takes and the cost of energy.
So what comfort does it actually get you?
- Your BC covers stay on the pipe
- Crafting signs and item signs stay on the pipe
- Any configuration or modules stay in the pipe
- You do not need to interrupt your network (item spilling or setting other pipes offline)
- You do not need to craft the next tier to replace your pipe

I honestly think that those points even overweight the cost of 30s and the cost of energy. I play Logistics Pipes for a few years now and I am confident that I will use this feature a lot. Especially when I have more than one pipe to upgrade, which happens later on in the game, when you want to replace many crafting pipes. This feature makes a lot of sense and is not overpriced, in my opinion.

Thanks for reading.

**Comment by TheLoneWolfling** [comment]
I believe @HaniiPuppy already covered the concerns I had in a much more eloquent manner than I ever could.

Suffice to say that adding time / resource costs for a convenience feature is a good way to prevent it from being used. If people are short on resources, they won't use it because of the resource cost, and if they aren't short on resources, they'll already be using the upgraded version even in places they don't need to because it saves time over the time cost of upgrading.

The other factor is that if I'm needing to upgrade a pipe it's often precisely because it's being overloaded - for example when the input ender quarry chest isn't getting pulled from fast enough. Making it go offline for an amount of time makes things worse.

**Comment by bziemons** [comment]
@TheLoneWolfling That is a matter of how you use Logistics Pipes. You cannot generally say that everyone will use Logistics Pipes that way and therefor the time penalty must be wrong. We know that not everyone uses Logistics Pipes the same way and therefor there are the config options.

**Comment by TheLoneWolfling** [comment]
It's a matter of how I, and everyone I've ever played on a server with LP with, uses LP. I am not saying it is wrong. I am saying that if it is implemented in the way you're saying it will probably will not be used much among the people I play with.

**Comment by AartBluestoke** [comment]
@HaniiPuppy "disable the pipe" only prevents items exiting at that point, or being requested from inventories attached to that pipe. items can still flow _through_ the pipe in the normal logistics manner.
Even more imporantly, there is no routing table rebuild where pipes on the left temporarily get seperated from pipes on the right - again routing _through_ the pipe is maintained. Upgrading the central part of a busy pipe without interrupting the flow from the upstream quarry is worth a fair bit of inconvenience in my mind.

**Comment by hron84** [comment]
Yeah, the biggest advantage of this feature against with break & replace is not breaks the LP system. I had many-many setups where everything went wrong if I replaced a single pipe especially when I used LP to transport a solid fuel to my power generation room. Also a big LP system can suffer a serious outage if the system has to be rebuilt every time when you upgrade pipes w/ break & replace. The most important idea behind this feature was keeping your whole LP system consistent and online during the upgrade - you just will not be able to use the extra feature of the pipe (crafting of crafting pipe, modules on chassis pipes, requesting on request pipes, etc.) until it gets done.

And I do not fear from being this feature unused. LP is not as popular as for example AE2 or XU so the target people are smart enough to see LP's advantage over these systems even if LP is a bit more complicated than the previously mentioned systems. These people would be happy if LP become a bit easier than before... :smile:
